### PR TITLE
Ivm

### PR DESCRIPTION
**Pull request by wesboyt**
Fixed IVM logic that corrupted the state of the lexer.
Fixed logic around invalidIVM badfiles.
Fixed null indexing/roundtripping bug(for some reason encountering null would set indexes to -1,-1.
Fixed writing out raw ion from event stream when testing failed.

### CODE REVIEW

File: src/IonParserTextRaw.ts [hunk 1]
```diff
@@ -561,6 +563,7 @@ export class ParserTextRaw {
                     symbol = this.get_value_as_string(T_IDENTIFIER);
                     kwt = get_type_from_name(symbol);
                 }
+                //this._start = tempNullStart;
```
Comment (tgregg) [inline src/IonParserTextRaw.ts]: Delete

File: src/IonTextReader.ts [hunk 2]
```diff
@@ -67,7 +69,7 @@ export class TextReader implements Reader {
   load_raw() {
     let t: TextReader = this;
     if (t._raw !== undefined) return;
-    if (t.isNull()) return;
+    //if (t.isNull()) return;
```
Comment (tgregg) [inline src/IonTextReader.ts]: Delete

File: src/IonTextReader.ts [hunk 3]
```diff
@@ -87,41 +89,62 @@ export class TextReader implements Reader {
...
-        this._symtab = makeSymbolTable(this._cat, this);
-      }
-      else {
-        break;
-      }
+        this._raw_type = p.next();
+        if (this._raw_type === T_IDENTIFIER) {
+            if (this._depth > 0) break;
+            this.load_raw();
+            //if(this._raw !== "$ion_1_0") break;
```
Comment (tgregg) [inline src/IonTextReader.ts]: Delete

File: src/IonTextReader.ts [hunk 4]
```diff
@@ -87,41 +89,62 @@ export class TextReader implements Reader {
...
+            if (!this.isIVM(this._raw)) break;
+            this._symtab = defaultLocalSymbolTable();
+            this._raw = undefined;
+            this._raw_type = undefined;
+        }else if (this._raw_type === T_STRING1) {
+            if (this._depth > 0) break;
+            this.load_raw();
+            if (this._raw !== "$ion_1_0") break;
```
Comment (tgregg) [inline src/IonTextReader.ts]: Is this needed?

**Comment by wesboyt** [comment]
The eventstream/event changes are from files that were taken off the skip list and revealed additional bugs since the parser was able to get past the IVM at the top now. the switch was falling through the case and producing extra ContainerEnd events. Unrelated to IVM specifically I agree.

File: src/IonTextReader.ts (hunk 4 above)
Comment (wesboyt) [reply src/IonTextReader.ts]: yes

Comment (tgregg) [reply src/IonTextReader.ts]: Oh because T_STRING1 mean's single-quoted symbol. Got it.

File: src/IonParserTextRaw.ts (hunk 1 above)
Comment (wesboyt) [reply src/IonParserTextRaw.ts]: done

File: src/IonTextReader.ts (hunk 2 above)
Comment (wesboyt) [reply src/IonTextReader.ts]: done

File: src/IonTextReader.ts (hunk 3 above)
Comment (wesboyt) [reply src/IonTextReader.ts]: done

**Pull request merged**